### ISSUE TITLE
Allow descriptions for `@template` tags

### ISSUE THREAD

**Issue by cherryblossom000** [gajus/eslint-plugin-jsdoc#653]
## Motivation

Allow descriptions for `@template` tags.

Sometimes, I need to document the type parameters of a type in TypeScript:

```ts
/**
 * Description of complicated type.
 *
 * @template T Description of the T type parameter.
 * @template U - Like other tags, this can have an optional hyphen before the description.
 */
type ComplicatedType<T, U> = // ...
```

Here's an example with a function from the [TypeScript handbook](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#template):

```js
/**
 * @template {string} K - K must be a string or string literal
 * @template {{ serious(): string }} Seriousalizable - must have a serious method
 * @param {K} key
 * @param {Seriousalizable} object
 */
function seriousalize(key, object) {
  // ????
}
```

## Current behavior

The `valid-types` rule reports errors like these:

```
Syntax error in namepath: T Description of the T type parameter.
Syntax error in namepath: K - K must be a string or string literal
```

## Desired behavior

Descriptions should be allowed for `@template` tags. Ideally, this should work for tags declaring multiple type parameters (`@template T,U`) as well.

```js
// Ok
/**
 * @template T Some description
 * @template U - With hyphen
 * @template V,W More parameters
 * @template W,X - Also with a hyphen
 */
function f() {}

// Invalid namepath for "invalid namepath" only
// Alternatively, make a hyphen mandatory to differentiate between the type parameters and the description.
/**
 * @template invalid namepath,T Description
 */
function f() {}
```

## Alternatives considered

Not allowing descriptions in `@template` tags or only allowing them in TypeScript mode because [GCC doesn't mention descriptions](https://github.com/google/closure-compiler/wiki/Generic-Types).

**Comment by brettz9** [comment]
Some good ideas... I'm a bit fatigued these days, but would accept a PR... I think allowing it only in TS mode for now would be the best option unless someone can confirm the Closure compiler works.

Can you determine in some way whether TS allows for descriptions without hyphens?

**Comment by cherryblossom000** [comment]
I put this in VSCode:

```ts
/**
 * @template T Without hyphen.
 * @template U - With hyphen.
 */
type SomeType<T, U> = never
```

Hovering over `SomeType` shows that the description for `T` (without the hyphen) is still recognised:

![screenshot of VSCode rendered JSDoc for SomeType](https://user-images.githubusercontent.com/31467609/97631277-d23ab480-1a84-11eb-8a53-6da774b9e2a0.png)

You can also try this out in the [TypeScript playground](https://www.typescriptlang.org/play?#code/PQKhCgAIUgBAXApgWwA4BsCGTIBVIDqAlvABYD2ArvJKQJ6qmIB2AdFDAihtopAKqQAtIRKlaDJmw7Bw8BnwDK5ZIlwKAPLgA0AgHyQAvJGaIAbogBO4IA).

Interestngly, TypeScript includes the hyphen in the `comment` property in the AST:

![screenshot of AST for JSDoc template tags](https://user-images.githubusercontent.com/31467609/97643744-a62b2d80-1a9c-11eb-8d24-2586c593417a.png)

**Comment by brettz9** [comment]
Ok, great, good info, thanks. Well, I think it's good for us in a sense if the hyphen is optional and part of the description, as that fits more with the approach of `comment-parser` (the parser we are normally using). However, our internal code does not use and probably cannot use `comment-parser` for the case of `@template` because [Closure](https://github.com/google/closure-compiler/wiki/Generic-Types#multiple-template-types) at least allows for whitespace between template names, which would otherwise signal the beginning of the description.

In source, the responsible function for parsing is the simple `parseClosureTemplateTag` in `jsdocUtils.js`.

It looks to me then that we need to begin the description where the first whitespace occurs that is not preceded by a comma.
